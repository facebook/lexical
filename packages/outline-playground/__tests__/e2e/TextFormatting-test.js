/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {
  initializeE2E,
  assertHTMLSnapshot,
  assertSelection,
  keyDownCtrlOrMeta,
  keyUpCtrlOrMeta,
  repeat,
  focusEditor,
} from '../utils';

describe('TextFormatting', () => {
  initializeE2E({chromium: true, webkit: true, firefox: true}, (e2e) => {
    e2e.flaky(['webkit'], () => {
      it(`Can create bold text using the shortcut`, async () => {
        const {page} = e2e;

        await focusEditor(page);
        await page.keyboard.type('Hello');
        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await page.keyboard.type(' World');
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 6,
          focusPath: [0, 1, 0],
          focusOffset: 6,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await page.keyboard.type('!');
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 2, 0],
          anchorOffset: 1,
          focusPath: [0, 2, 0],
          focusOffset: 1,
        });
      });

      it(`Can create italic text using the shortcut`, async () => {
        const {page} = e2e;

        await focusEditor(page);
        await page.keyboard.type('Hello');
        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await page.keyboard.type(' World');
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 6,
          focusPath: [0, 1, 0],
          focusOffset: 6,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await page.keyboard.type('!');
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 2, 0],
          anchorOffset: 1,
          focusPath: [0, 2, 0],
          focusOffset: 1,
        });
      });

      it(`Can select text and boldify it with the shortcut`, async () => {
        const {page} = e2e;

        await focusEditor(page);
        await page.keyboard.type('Hello world!');
        await page.keyboard.press('ArrowLeft');
        await page.keyboard.down('Shift');
        await repeat(5, async () => {
          await page.keyboard.press('ArrowLeft');
        });
        await page.keyboard.up('Shift');
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 11,
          focusPath: [0, 0, 0],
          focusOffset: 6,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 0,
          focusPath: [0, 1, 0],
          focusOffset: 5,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 6,
          focusPath: [0, 0, 0],
          focusOffset: 11,
        });
      });

      it(`Can select text and italicify it with the shortcut`, async () => {
        const {page} = e2e;

        await focusEditor(page);
        await page.keyboard.type('Hello world!');
        await page.keyboard.press('ArrowLeft');
        await page.keyboard.down('Shift');
        await repeat(5, async () => {
          await page.keyboard.press('ArrowLeft');
        });
        await page.keyboard.up('Shift');
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 11,
          focusPath: [0, 0, 0],
          focusOffset: 6,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 0,
          focusPath: [0, 1, 0],
          focusOffset: 5,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 6,
          focusPath: [0, 0, 0],
          focusOffset: 11,
        });
      });

      it(`Can select multiple text parts and format them with shortcuts`, async () => {
        const {page} = e2e;

        await focusEditor(page);
        await page.keyboard.type('Hello world!');
        await page.keyboard.press('ArrowLeft');
        await page.keyboard.down('Shift');
        await repeat(5, async () => {
          await page.keyboard.press('ArrowLeft');
        });
        await page.keyboard.up('Shift');
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 11,
          focusPath: [0, 0, 0],
          focusOffset: 6,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 0,
          focusPath: [0, 1, 0],
          focusOffset: 5,
        });

        await page.keyboard.press('ArrowLeft');
        await page.keyboard.press('ArrowRight');
        await page.keyboard.down('Shift');
        await page.keyboard.press('ArrowRight');
        await page.keyboard.press('ArrowRight');
        await page.keyboard.up('Shift');
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 1,
          focusPath: [0, 1, 0],
          focusOffset: 3,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 2, 0],
          anchorOffset: 0,
          focusPath: [0, 2, 0],
          focusOffset: 2,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 2, 0],
          anchorOffset: 0,
          focusPath: [0, 2, 0],
          focusOffset: 2,
        });

        await page.keyboard.press('ArrowLeft');
        await page.keyboard.press('ArrowLeft');
        await page.keyboard.down('Shift');
        await repeat(5, async () => {
          await page.keyboard.press('ArrowRight');
        });
        await page.keyboard.up('Shift');
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 0,
          focusPath: [0, 3, 0],
          focusOffset: 2,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('b');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 6,
          focusPath: [0, 2, 0],
          focusOffset: 2,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 1, 0],
          anchorOffset: 0,
          focusPath: [0, 1, 0],
          focusOffset: 5,
        });

        await keyDownCtrlOrMeta(page);
        await page.keyboard.press('i');
        await keyUpCtrlOrMeta(page);
        await assertHTMLSnapshot(page);
        await assertSelection(page, {
          anchorPath: [0, 0, 0],
          anchorOffset: 6,
          focusPath: [0, 0, 0],
          focusOffset: 11,
        });
      });
    });
  });
});
