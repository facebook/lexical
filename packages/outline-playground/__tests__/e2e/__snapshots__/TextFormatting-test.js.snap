// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFormatting Can create bold text using the shortcut 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello
  </span>
  <strong
    class="editor-text-bold"
  >
     World
  </strong>
</p>
`;

exports[`TextFormatting Can create bold text using the shortcut 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello
  </span>
  <strong
    class="editor-text-bold"
  >
     World
  </strong>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can create italic text using the shortcut 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello
  </span>
  <em
    class="editor-text-italic"
  >
     World
  </em>
</p>
`;

exports[`TextFormatting Can create italic text using the shortcut 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello
  </span>
  <em
    class="editor-text-italic"
  >
     World
  </em>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <strong
    class="editor-text-bold"
  >
    world
  </strong>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <strong
    class="editor-text-bold"
  >
    w
  </strong>
  <strong
    class="editor-text-bold editor-text-italic"
  >
    or
  </strong>
  <strong
    class="editor-text-bold"
  >
    ld
  </strong>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 3`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <strong
    class="editor-text-bold"
  >
    w
  </strong>
  <em
    class="editor-text-italic"
  >
    or
  </em>
  <strong
    class="editor-text-bold"
  >
    ld
  </strong>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 4`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello w
  </span>
  <em
    class="editor-text-italic"
  >
    or
  </em>
  <span>
    ld!
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 5`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <em
    class="editor-text-italic"
  >
    world
  </em>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select multiple text parts and format them with shortcuts 6`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello world!
  </span>
</p>
`;

exports[`TextFormatting Can select text and boldify it with the shortcut 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <strong
    class="editor-text-bold"
  >
    world
  </strong>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select text and boldify it with the shortcut 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello world!
  </span>
</p>
`;

exports[`TextFormatting Can select text and italicify it with the shortcut 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello 
  </span>
  <em
    class="editor-text-italic"
  >
    world
  </em>
  <span>
    !
  </span>
</p>
`;

exports[`TextFormatting Can select text and italicify it with the shortcut 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    Hello world!
  </span>
</p>
`;
